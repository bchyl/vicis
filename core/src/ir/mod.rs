pub mod function;
pub mod module;
pub mod types;
pub mod util;
pub mod value;
