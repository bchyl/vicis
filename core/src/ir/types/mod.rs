pub mod parser;

use crate::ir::module::name::Name;
use id_arena::{Arena, Id};
use rustc_hash::FxHashMap;
use std::{
    cell::{Ref, RefCell, RefMut},
    fmt, mem,
    sync::Arc,
};

pub use parser::parse;

pub type AddrSpace = u32;
pub type Cache<T> = FxHashMap<T, TypeId>;
pub type TypeId = Id<Type>;

#[derive(Clone)]
pub struct Types(Arc<RefCell<TypesBase>>);

pub struct TypesBase {
    arena: Arena<Type>,
    named_types: Cache<Name>,
    void: TypeId,
    int: Cache<u32>,
    pointer: Cache<(TypeId, u32)>,
    array: Cache<(TypeId, u32)>,
    structs: Cache<String>,
    metadata: TypeId,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Type {
    Void,
    Int(u32),
    Pointer(PointerType),
    Array(ArrayType),
    Function(FunctionType),
    Struct(StructType),
    Metadata,
    // TODO: Add more types
}

#[derive(Debug, PartialEq, Eq, Hash, Clone)]
pub struct PointerType {
    pub inner: TypeId,
    pub addr_space: AddrSpace,
}

#[derive(Debug, PartialEq, Eq, Hash, Clone)]
pub struct ArrayType {
    pub inner: TypeId,
    pub num_elements: u32,
}

#[derive(Debug, PartialEq, Eq, Clone, Hash)]
pub struct FunctionType {
    pub ret: TypeId,
    pub params: Vec<TypeId>,
    pub is_var_arg: bool,
}

#[derive(Debug, PartialEq, Eq, Clone, Hash)]
pub struct StructType {
    pub name: Option<Name>,
    pub elems: Vec<TypeId>,
    pub is_packed: bool,
}

impl Default for Types {
    fn default() -> Self {
        Self(Arc::new(RefCell::new(TypesBase::new())))
    }
}

impl Types {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn base(&self) -> Ref<TypesBase> {
        self.0.borrow()
    }

    pub fn base_mut(&self) -> RefMut<TypesBase> {
        self.0.borrow_mut()
    }

    pub fn get(&self, id: TypeId) -> Ref<Type> {
        Ref::map(self.0.borrow(), |base| &base.arena[id])
    }

    pub fn get_mut(&self, id: TypeId) -> RefMut<Type> {
        RefMut::map(self.0.borrow_mut(), |base| &mut base.arena[id])
    }

    // Get element type of pointer or array type
    pub fn get_element(&self, id: TypeId) -> Option<TypeId> {
        self.0.borrow().element(id)
    }

    pub fn is_atomic(&self, id: TypeId) -> bool {
        self.base().is_atomic(id)
    }

    pub fn to_string(&self, ty: TypeId) -> String {
        self.base().to_string(ty)
    }

    // TODO: We should add a trait to implement methods below.

    pub fn void(&self) -> TypeId {
        self.base().void()
    }

    pub fn i1(&self) -> TypeId {
        self.base().i1()
    }

    pub fn i8(&self) -> TypeId {
        self.base().i8()
    }

    pub fn i16(&self) -> TypeId {
        self.base().i16()
    }

    pub fn i32(&self) -> TypeId {
        self.base().i32()
    }

    pub fn i64(&self) -> TypeId {
        self.base().i64()
    }

    pub fn metadata(&self) -> TypeId {
        self.base().metadata()
    }
}

impl Default for TypesBase {
    fn default() -> Self {
        let mut arena = Arena::new();
        let void = arena.alloc(Type::Void);
        let int: Cache<u32> = vec![
            (1, Type::Int(1)),
            (8, Type::Int(8)),
            (16, Type::Int(16)),
            (32, Type::Int(32)),
            (64, Type::Int(64)),
        ]
        .into_iter()
        .map(|(bits, ty)| (bits, arena.alloc(ty)))
        .collect();
        let metadata = arena.alloc(Type::Metadata);
        Self {
            arena,
            named_types: Cache::default(),
            void,
            int,
            pointer: Cache::default(),
            array: Cache::default(),
            structs: Cache::default(),
            metadata,
        }
    }
}

impl TypesBase {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn named_type(&mut self, name: Name) -> TypeId {
        if let Some(named) = self.named_types.get(&name) {
            return *named;
        }
        let id = self.arena.alloc(Type::Void);
        self.named_types.insert(name, id);
        id
    }

    pub fn void(&self) -> TypeId {
        self.void
    }

    pub fn int(&mut self, bits: u32) -> TypeId {
        if let Some(int) = self.int.get(&bits) {
            return *int;
        }
        let id = self.arena.alloc(Type::Int(bits));
        self.int.insert(bits, id);
        id
    }

    pub fn i1(&self) -> TypeId {
        self.int[&1]
    }

    pub fn i8(&self) -> TypeId {
        self.int[&8]
    }

    pub fn i16(&self) -> TypeId {
        self.int[&16]
    }

    pub fn i32(&self) -> TypeId {
        self.int[&32]
    }

    pub fn i64(&self) -> TypeId {
        self.int[&64]
    }

    pub fn metadata(&self) -> TypeId {
        self.metadata
    }

    pub fn pointer(&mut self, inner: TypeId) -> TypeId {
        if let Some(pointer) = self.pointer.get(&(inner, 0)) {
            return *pointer;
        }
        let id = self.arena.alloc(Type::Pointer(PointerType {
            inner,
            addr_space: 0,
        }));
        self.pointer.insert((inner, 0), id);
        id
    }

    pub fn pointer_in_addr_space(&mut self, inner: TypeId, addr_space: u32) -> TypeId {
        if let Some(pointer) = self.pointer.get(&(inner, addr_space)) {
            return *pointer;
        }
        let id = self
            .arena
            .alloc(Type::Pointer(PointerType { inner, addr_space }));
        self.pointer.insert((inner, addr_space), id);
        id
    }

    pub fn array(&mut self, inner: TypeId, num_elements: u32) -> TypeId {
        if let Some(array) = self.array.get(&(inner, num_elements)) {
            return *array;
        }
        let id = self.arena.alloc(Type::Array(ArrayType {
            inner,
            num_elements,
        }));
        self.array.insert((inner, num_elements), id);
        id
    }

    pub fn function(&mut self, ret: TypeId, params: Vec<TypeId>, is_var_arg: bool) -> TypeId {
        // TODO: FIXME: Should cache function type?
        self.arena.alloc(Type::Function(FunctionType {
            ret,
            params,
            is_var_arg,
        }))
    }

    pub fn empty_struct_named(&mut self, name: String, is_packed: bool) -> TypeId {
        if let Some(strct) = self.structs.get(&name) {
            return *strct;
        }
        let id = self.arena.alloc(Type::Struct(StructType {
            name: Some(Name::Name(name.clone())),
            elems: vec![],
            is_packed,
        }));
        self.structs.insert(name, id);
        id
    }

    pub fn anonymous_struct(&mut self, elems: Vec<TypeId>, is_packed: bool) -> TypeId {
        self.arena.alloc(Type::Struct(StructType {
            name: None,
            elems,
            is_packed,
        }))
    }

    pub fn get_struct(&self, name: &str) -> Option<TypeId> {
        self.structs.get(name).copied()
    }

    pub fn change_to_named_type(&mut self, ty: TypeId, name: Name) {
        let named_ty = self.named_type(name.clone());

        if self.is_struct(ty) {
            let mut strukt = mem::replace(&mut self.arena[ty], Type::Void);
            if let Some(name) = name.to_string() {
                strukt.as_struct_mut().name = Some(Name::Name(name.to_owned()));
                self.structs.insert(name.to_owned(), named_ty);
            }
            self.arena[named_ty] = strukt;
            return;
        }

        let ty = self.arena[ty].clone();
        self.arena[named_ty] = ty;
    }

    pub fn element(&self, ty: TypeId) -> Option<TypeId> {
        match self.arena[ty] {
            Type::Void => None,
            Type::Int(_) => None,
            Type::Pointer(PointerType { inner, .. }) => Some(inner),
            Type::Array(ArrayType { inner, .. }) => Some(inner),
            Type::Function(_) => None,
            Type::Struct(_) => None,
            Type::Metadata => None,
        }
    }

    pub fn element_at(&self, ty: TypeId, i: usize) -> Option<TypeId> {
        match self.arena[ty] {
            Type::Void => None,
            Type::Int(_) => None,
            Type::Pointer(_) => None,
            Type::Array(ArrayType { inner, .. }) => Some(inner),
            Type::Function(_) => None,
            Type::Struct(StructType { ref elems, .. }) => elems.get(i).copied(),
            Type::Metadata => None,
        }
    }

    pub fn to_string(&self, ty: TypeId) -> String {
        let ty = &self.arena[ty];
        match ty {
            Type::Void => "void".to_string(),
            Type::Int(bits) => format!("i{}", bits),
            Type::Pointer(PointerType { inner, addr_space }) if *addr_space == 0 => {
                format!("{}*", self.to_string(*inner))
            }
            Type::Pointer(PointerType { inner, addr_space }) => {
                format!("{} addrspace({})*", self.to_string(*inner), addr_space)
            }
            Type::Array(ArrayType {
                inner,
                num_elements,
            }) => {
                format!("[{} x {}]", num_elements, self.to_string(*inner))
            }
            Type::Function(FunctionType {
                ret,
                params,
                is_var_arg,
            }) => {
                format!(
                    "{} ({})",
                    self.to_string(*ret),
                    params
                        .iter()
                        .enumerate()
                        .fold("".to_string(), |acc, (i, &param)| {
                            format!(
                                "{}{}{}",
                                acc,
                                self.to_string(param),
                                if i == params.len() - 1 {
                                    if *is_var_arg {
                                        ", ..."
                                    } else {
                                        ""
                                    }
                                } else {
                                    ", "
                                }
                            )
                        })
                )
            }
            Type::Struct(ty) => {
                if let Some(name) = ty.name.as_ref() {
                    return format!("%{}", name);
                }
                self.struct_definition_to_string(ty)
            }
            Type::Metadata => "metadata".to_string(),
        }
    }

    pub fn struct_definition_to_string(&self, ty: &StructType) -> String {
        let mut elems_str = "".to_string();
        for (i, elem) in ty.elems.iter().enumerate() {
            elems_str.push_str(self.to_string(*elem).as_str());
            if i != ty.elems.len() - 1 {
                elems_str.push_str(", ");
            }
        }
        format!(
            "{}{{ {} }}{}",
            if ty.is_packed { "<" } else { "" },
            elems_str,
            if ty.is_packed { ">" } else { "" }
        )
    }

    pub fn is_struct(&self, ty: TypeId) -> bool {
        matches!(self.arena[ty], Type::Struct(_))
    }

    pub fn is_atomic(&self, ty: TypeId) -> bool {
        let ty = &self.arena[ty];
        matches!(ty, Type::Void | Type::Pointer(_) | Type::Int(_))
    }
}

impl Type {
    pub fn as_struct(&self) -> &StructType {
        match self {
            Self::Struct(strct) => strct,
            _ => panic!(),
        }
    }

    pub fn as_struct_mut(&mut self) -> &mut StructType {
        match self {
            Self::Struct(strct) => strct,
            _ => panic!(),
        }
    }
}

impl fmt::Debug for Types {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for (name, &id) in &self.base().named_types {
            writeln!(
                f,
                "%{} = type {}",
                name,
                match &*self.get(id) {
                    Type::Struct(ty) => {
                        self.base().struct_definition_to_string(ty)
                    }
                    _ => self.to_string(id),
                }
            )?
        }
        Ok(())
    }
}

#[test]
fn types_identity() {
    let types = Types::new();
    let i32_ptr_ty = {
        let i32_ty = types.i32();
        types.base_mut().pointer(i32_ty)
    };

    {
        let i32_ty = types.i32();
        let ty = types.get(i32_ptr_ty);
        assert_eq!(
            &*ty,
            &Type::Pointer(PointerType {
                inner: i32_ty,
                addr_space: 0
            })
        )
    }

    let i32_ty = types.i32();
    let i32_ptr_ty2 = types.base_mut().pointer(i32_ty);

    assert_eq!(i32_ptr_ty, i32_ptr_ty2);
}
