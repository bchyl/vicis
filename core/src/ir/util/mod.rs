use std::{collections::VecDeque, str};

pub fn unescape(s: &str) -> Option<String> {
    let mut queue: VecDeque<_> = String::from(s).chars().collect();
    let mut s = String::new();

    while let Some(c) = queue.pop_front() {
        if c != '\\' {
            s.push(c);
            continue;
        }

        match queue.pop_front() {
            Some('\\') => s.push('\\'),
            Some(c) if c.is_ascii_hexdigit() => {
                let cc = queue.pop_front().unwrap();
                s.push(
                    char::from_u32(
                        u32::from_str_radix(format!("{}{}", c, cc).as_str(), 16).unwrap(),
                    )
                    .unwrap(),
                );
            }
            _ => return None,
        };
    }

    Some(s)
}

pub fn escape(s: &str) -> String {
    return str::from_utf8(
        &s.as_bytes()
            .iter()
            .flat_map(|&c| {
                if c.is_ascii_control() {
                    vec![b'\\', hexify(c >> 4), hexify(c & 0xf)]
                } else {
                    vec![c]
                }
                .into_iter()
            })
            .collect::<Vec<u8>>(),
    )
    .unwrap()
    .to_string();

    fn hexify(b: u8) -> u8 {
        match b {
            0..=9 => b'0' + b,
            _ => b'a' + b - 10,
        }
    }
}
