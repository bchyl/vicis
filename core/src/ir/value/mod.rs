pub mod parser;

pub use parser::parse;

use super::{
    function::{data::Data, instruction::InstructionId},
    module::name::Name,
    types::{TypeId, Types},
    util::escape,
};
use id_arena::Id;
use std::{fmt, str};

pub type ValueId = Id<Value>;

#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Instruction(InstructionId),
    Argument(usize),
    Constant(ConstantData),
    InlineAsm(InlineAsm),
}

#[derive(Debug, Clone, PartialEq)]
pub enum ConstantData {
    Undef,
    AggregateZero,
    Null,
    Int(ConstantInt),
    Array(ConstantArray),
    Struct(ConstantStruct),
    Expr(ConstantExpr), // TODO: Boxing?
    GlobalRef(Name),
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ConstantInt {
    Int1(bool),
    Int8(i8),
    Int32(i32),
    Int64(i64),
}

#[derive(Debug, Clone, PartialEq)]
pub struct ConstantArray {
    pub elem_ty: TypeId,
    pub elems: Vec<ConstantData>,
    pub is_string: bool,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ConstantStruct {
    pub elems_ty: Vec<TypeId>,
    pub elems: Vec<ConstantData>,
    pub is_packed: bool,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ConstantExpr {
    GetElementPtr {
        inbounds: bool,
        tys: Vec<TypeId>,
        args: Vec<ConstantData>,
    },
    Bitcast {
        tys: [TypeId; 2],
        arg: Box<ConstantData>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct InlineAsm {
    pub body: String,
    pub constraints: String,
    pub sideeffect: bool,
}

impl Value {
    pub fn undef() -> Self {
        Self::Constant(ConstantData::Undef)
    }

    pub fn as_inst(&self) -> &InstructionId {
        match self {
            Self::Instruction(id) => id,
            _ => panic!(),
        }
    }

    pub fn to_string(&self, _data: &Data, types: &Types) -> String {
        match self {
            Self::Constant(c) => c.to_string(types),
            Self::Instruction(id) => {
                format!("%I{}", id.index())
            }
            Self::Argument(n) => format!("%A{}", n),
            Self::InlineAsm(InlineAsm {
                body,
                constraints,
                sideeffect,
            }) => {
                format!(
                    "asm {}\"{}\", \"{}\"",
                    if *sideeffect { "sideeffect " } else { "" },
                    constraints,
                    body
                )
            }
        }
    }
}

impl ConstantData {
    pub fn to_string(&self, types: &Types) -> String {
        match self {
            Self::Undef => "undef".to_string(),
            Self::AggregateZero => "AggregateZero".to_string(),
            Self::Null => "null".to_string(),
            Self::Int(i) => i.to_string(),
            Self::Array(a) => a.to_string(types),
            Self::Struct(s) => s.to_string(types),
            Self::Expr(e) => e.to_string(types),
            Self::GlobalRef(name) => format!("@{:?}", name),
        }
    }

    pub fn as_int(&self) -> &ConstantInt {
        match self {
            Self::Int(i) => i,
            _ => panic!(),
        }
    }

    pub fn as_global_ref(&self) -> &Name {
        match self {
            Self::GlobalRef(name) => name,
            _ => panic!(),
        }
    }

    pub fn as_array(&self) -> &ConstantArray {
        match self {
            Self::Array(a) => a,
            _ => panic!(),
        }
    }
}

impl ConstantInt {
    pub fn as_i8(&self) -> &i8 {
        match self {
            Self::Int8(i) => i,
            _ => panic!(),
        }
    }

    // TODO: Generics?
    pub fn cast_to_usize(self) -> usize {
        match self {
            Self::Int1(i) => i as usize,
            Self::Int8(i) => i as usize,
            Self::Int32(i) => i as usize,
            Self::Int64(i) => i as usize,
        }
    }
}

impl ConstantArray {
    pub fn to_string(&self, types: &Types) -> String {
        if self.is_string {
            return format!(
                "c\"{}\"",
                escape(
                    str::from_utf8(
                        self.elems
                            .iter()
                            .map(|i| *i.as_int().as_i8() as u8)
                            .collect::<Vec<u8>>()
                            .as_slice(),
                    )
                    .unwrap()
                )
            );
        }

        format!(
            "{}]",
            self.elems
                .iter()
                .fold("[".to_string(), |acc, e| {
                    format!(
                        "{}{} {}, ",
                        acc,
                        types.to_string(self.elem_ty),
                        e.to_string(types)
                    )
                })
                .trim_end_matches(", ")
        )
    }
}

impl ConstantStruct {
    pub fn to_string(&self, types: &Types) -> String {
        format!(
            "{}{{ {} }}{}",
            if self.is_packed { "<" } else { "" },
            self.elems
                .iter()
                .zip(self.elems_ty.iter())
                .fold("".to_string(), |acc, (e, &et)| {
                    format!("{}{} {}, ", acc, types.to_string(et), e.to_string(types))
                })
                .trim_end_matches(", "),
            if self.is_packed { ">" } else { "" }
        )
    }
}

impl ConstantExpr {
    pub fn to_string(&self, types: &Types) -> String {
        match self {
            Self::GetElementPtr {
                inbounds,
                tys,
                args,
            } => {
                format!(
                    "getelementptr {}({}, {})",
                    if *inbounds { "inbounds " } else { "" },
                    types.to_string(tys[0]),
                    tys[1..]
                        .iter()
                        .zip(args.iter())
                        .fold("".to_string(), |acc, (ty, arg)| {
                            format!("{}{} {}, ", acc, types.to_string(*ty), arg.to_string(types))
                        })
                        .trim_end_matches(", ")
                )
            }
            Self::Bitcast { tys, arg } => {
                format!(
                    "bitcast ({} {} to {})",
                    types.to_string(tys[0]),
                    arg.to_string(types),
                    types.to_string(tys[1]),
                )
            }
        }
    }
}

impl From<i32> for Value {
    fn from(i: i32) -> Self {
        Self::Constant(ConstantInt::Int32(i).into())
    }
}

impl From<i64> for Value {
    fn from(i: i64) -> Self {
        Self::Constant(ConstantInt::Int64(i).into())
    }
}

impl From<ConstantInt> for ConstantData {
    fn from(i: ConstantInt) -> Self {
        Self::Int(i)
    }
}

impl From<ConstantInt> for Value {
    fn from(i: ConstantInt) -> Self {
        Self::Constant(i.into())
    }
}

impl fmt::Display for ConstantInt {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Int1(i) => write!(f, "{}", i),
            Self::Int8(i) => write!(f, "{}", i),
            Self::Int32(i) => write!(f, "{}", i),
            Self::Int64(i) => write!(f, "{}", i),
        }
    }
}
