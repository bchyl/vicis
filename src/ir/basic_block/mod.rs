use super::module::name::Name;
use id_arena::Id;
use rustc_hash::FxHashSet;

pub type BasicBlockId = Id<BasicBlock>;

pub struct BasicBlock {
    pub name: Option<Name>,
    pub preds: FxHashSet<BasicBlockId>,
    pub succs: FxHashSet<BasicBlockId>,
}

impl BasicBlock {
    pub fn new() -> Self {
        Self {
            name: None,
            preds: FxHashSet::default(),
            succs: FxHashSet::default(),
        }
    }

    pub fn preds(&self) -> &FxHashSet<BasicBlockId> {
        &self.preds
    }

    pub fn preds_mut(&mut self) -> &mut FxHashSet<BasicBlockId> {
        &mut self.preds
    }

    pub fn succs(&self) -> &FxHashSet<BasicBlockId> {
        &self.succs
    }

    pub fn succs_mut(&mut self) -> &mut FxHashSet<BasicBlockId> {
        &mut self.succs
    }
}
