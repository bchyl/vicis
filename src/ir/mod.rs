pub mod function;
pub mod instruction;
pub mod module;
pub mod types;
pub mod util;
pub mod value;
