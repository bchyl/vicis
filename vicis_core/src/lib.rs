#![feature(vec_into_raw_parts)]

// pub mod codegen;
// pub mod exec;
#[macro_use]
pub mod macros;
pub mod ir;
pub mod pass;
pub mod traits;

extern crate anyhow;
extern crate nom;
