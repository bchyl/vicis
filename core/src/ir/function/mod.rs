pub mod basic_block;
pub mod builder;
pub mod data;
pub mod instruction;
pub mod layout;
pub mod param_attrs;
pub mod print;

use super::{
    module::{
        attributes::Attribute, linkage::Linkage, name::Name,
        preemption_specifier::PreemptionSpecifier, unnamed_addr::UnnamedAddr,
        visibility::Visibility,
    },
    types::{Type, Types},
    value::ConstantData,
};
use crate::traits::basic_block::{BasicBlockData, BasicBlockLayout};
use basic_block::BasicBlock;
use id_arena::Id;
use instruction::InstructionId;
use param_attrs::ParameterAttribute;
use std::fmt;

pub type FunctionId = Id<Function>;

pub type PersonalityFunc = (Type, ConstantData);

pub struct Function {
    pub name: String,
    pub is_var_arg: bool,
    pub result_ty: Type,
    pub params: Vec<Parameter>,
    pub linkage: Linkage,
    pub preemption_specifier: PreemptionSpecifier,
    pub visibility: Visibility,
    pub unnamed_addr: Option<UnnamedAddr>,
    pub func_attrs: Vec<Attribute>,
    pub ret_attrs: Vec<param_attrs::ParameterAttribute>,
    pub personality: Option<PersonalityFunc>,
    pub data: data::Data,
    pub layout: layout::Layout,
    pub types: Types,
    // pub is_prototype: bool,
}

#[derive(Debug, Clone)]
pub struct Parameter {
    pub name: Name,
    pub ty: Type,
    pub attrs: Vec<ParameterAttribute>,
}

impl Function {
    pub fn new<T: AsRef<str>>(
        name: T,
        result_ty: Type,
        params: Vec<Parameter>,
        is_var_arg: bool,
        types: Types,
    ) -> Self {
        Self {
            name: name.as_ref().to_string(),
            is_var_arg,
            result_ty,
            params,
            linkage: Linkage::Common,
            preemption_specifier: PreemptionSpecifier::DsoLocal,
            visibility: Visibility::Default,
            unnamed_addr: None,
            func_attrs: vec![],
            ret_attrs: vec![],
            personality: None,
            data: data::Data::default(),
            layout: layout::Layout::default(),
            types,
        }
    }

    pub fn name(&self) -> &String {
        &self.name
    }

    pub fn params(&self) -> &[Parameter] {
        &self.params
    }

    pub fn is_var_arg(&self) -> bool {
        self.is_var_arg
    }

    pub fn is_prototype(&self) -> bool {
        self.layout.is_empty()
    }

    pub fn remove_inst(&mut self, inst: InstructionId) -> Option<()> {
        self.data.remove_uses(inst);
        self.layout.remove_inst(inst)
    }
}

impl fmt::Debug for Function {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        print::FunctionAsmPrinter::new(f).print(self)
    }
}

impl Parameter {
    pub fn new(ty: Type) -> Self {
        Self {
            name: Name::Number(0),
            ty,
            attrs: vec![],
        }
    }

    pub fn to_string(&self, types: &Types) -> String {
        format!("{} %{:?}", types.to_string(self.ty), self.name)
    }
}

impl BasicBlockData<BasicBlock> for Function {
    fn get(&self, id: Id<BasicBlock>) -> &BasicBlock {
        &self.data.basic_blocks[id]
    }
}

impl BasicBlockLayout<BasicBlock> for Function {
    fn order(&self) -> Box<dyn Iterator<Item = Id<BasicBlock>> + '_> {
        Box::new(self.layout.block_iter())
    }
}
