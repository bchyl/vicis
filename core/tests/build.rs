use vicis_core::ir::{function, module};

#[test]
fn build() {
    let mut module = module::Module::default();
    let int = module.types.i32();
    let func = module.create_function("func", int, vec![], false);
    let func = &mut module.functions_mut()[func];
    let mut builder = function::builder::Builder::new(func);
    let entry = builder.create_block();
    builder.switch_to_block(entry);
    let forty_two = builder.value(42i32);
    builder.inst().ret(forty_two);

    insta::assert_debug_snapshot!(module);
}
