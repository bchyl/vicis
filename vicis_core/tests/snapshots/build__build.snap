---
source: tests/build.rs
expression: module

---
source_filename = ""
target datalayout = ""
target triple = ""


define common dso_local default i32 @func() {
0:
    ret i32 42
}


